File: 2054-the-number-of-the-smallest-unoccupied-chair/the-number-of-the-smallest-unoccupied-chair.py
class Solution:
    def smallestChair(self, times: List[List[int]], targetFriend: int) -> int:
        #  Hashmap? 0 .. n-1 # Friend:Index
        hm = {i:tup for i, tup in enumerate(times)}
        shm = sorted(hm.items(), key = lambda x: x[1][0])
       
    
        occupied = []
        minUnoccupied = []
        nextChair = 0

        for friend, time in shm:
            arr, leav = time

            while occupied and arr >= occupied[0][0]:
                _, chairNum = heapq.heappop(occupied)
                heapq.heappush(minUnoccupied, chairNum)

            if minUnoccupied:
                chairNum = heapq.heappop(minUnoccupied)
            else:
                chairNum = nextChair
                nextChair = chairNum + 1
                
            heapq.heappush(occupied, (leav, chairNum))
    
            if friend == targetFriend:
                return chairNum
        return -1


          

File: 2054-the-number-of-the-smallest-unoccupied-chair/the-number-of-the-smallest-unoccupied-chair.java
class Solution {
    public int smallestChair(int[][] times, int targetFriend) {
        // int n = times.length;
        // int[][] friends = new int[n][3];

        // for(int i= 0; i < n ; i++){
        //     friends[i] = new int[]{times[i][0], times[i][1], i};
        // }
        // Arrays.sort(friends, (a,b) -> a[0]-b[0]);

        int targetArrival = times[targetFriend][0];
        Arrays.sort(times, (a,b) -> a[0]-b[0]);
        
        PriorityQueue<Integer> unOccupied = new PriorityQueue<>();
        PriorityQueue<int[]> Occupied = new PriorityQueue<>((a,b) -> a[0] -b[0]);
        int nextChair = 0;
        int chair = 0;

        // System.out.println(Arrays.deepToString(friends));
        for(int[] time: times) {

            int arr = time[0], leav = time[1];
            
            while(!Occupied.isEmpty() && Occupied.peek()[0] <= arr){
                chair = Occupied.poll()[1];   
                unOccupied.offer(chair);
            }

            chair  = unOccupied.isEmpty() ? nextChair++ : unOccupied.poll();

            // if (unOccupied.isEmpty()) {
            //     chair = nextChair;
            //     nextChair = chair + 1;
            // }
            // else {
            //     chair = unOccupied.poll();
            // }
            Occupied.offer(new int[] {leav, chair}); 

            if(targetArrival == arr){
                return chair;
            }

        }
        return 0;
    }
}